### PR TITLE
MsalBrowserAuthenticationProvider 

### LINKED ISSUE #483: #373 Part 1 - Add support for Msal Browser

### PR DESCRIPTION
**Pull request by nikithauc**
Resolves #483.

### CODE REVIEW

File: docs/MSALBrowserAuthenticationProvider.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+Steps to use `MSALBrowserAuthenticationProvider`;
+
+1.  Using npm: `npm install @azure/msal-browser`
+
+    Using html:
+
+    ```html
+    <!--Using script tag to include the bundled file or the CDN source-->
+    <script type="text/javascript" src="https://alcdn.msauth.net/browser/2.15.0/js/msal-browser.min.js"></script>
```
Comment (baywet) [inline docs/MSALBrowserAuthenticationProvider.md:18]: use of single tags? 
integrity and cross origin attributes?

File: package.json [hunk 2]
```diff
@@ -1,135 +1,139 @@
...
+        },
+        "buffer": {
+            "optional": true
+        },
+        "msal": {
+            "optional": true
+        },
+        "stream-browserify": {
+            "optional": true
+        },
```
Comment (baywet) [inline package.json:131]: remove msal above?

File: src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	 * Creates an instance of ImplicitMSALAuthenticationProvider
+	 * @param {PublicClientApplication} msalApplication - An instance of MSAL PublicClientApplication
+	 * @param {MSALBrowserAuthenticationProviderOptions} options - An instance of MSALAuthenticationProviderOptions
+	 * @returns An instance of ImplicitMSALAuthenticationProvider
+	 */
+	public constructor(private publicClientApplication: PublicClientApplication, private options: MSALBrowserAuthenticationProviderOptions) {
+		if (!this.options || !this.publicClientApplication) {
+			throw new GraphClientError("Please pass valid PublicClientApplication instance and MSALBrowserAuthenticationProviderOptions instance to instantiate MSALBrowserAuthenticationProvider");
+		}
+		this.options = options;
```
Comment (baywet) [inline src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts:37]: you shouldn't need to assign things, the transpiled code should do it for you.

File: src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	/**
+	 * @public
+	 * @constructor
+	 * Creates an instance of ImplicitMSALAuthenticationProvider
+	 * @param {PublicClientApplication} msalApplication - An instance of MSAL PublicClientApplication
+	 * @param {MSALBrowserAuthenticationProviderOptions} options - An instance of MSALAuthenticationProviderOptions
+	 * @returns An instance of ImplicitMSALAuthenticationProvider
+	 */
+	public constructor(private publicClientApplication: PublicClientApplication, private options: MSALBrowserAuthenticationProviderOptions) {
+		if (!this.options || !this.publicClientApplication) {
```
Comment (baywet) [inline src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts:34]:
```suggestion
		if (!options || !publicClientApplication) {
```

File: src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			});
+			return response.accessToken;
+		} catch (error) {
+			if (error instanceof InteractionRequiredAuthError) {
+				if (this.options.interactionType === InteractionType.Redirect) {
+					this.publicClientApplication.acquireTokenRedirect({ scopes });
+				} else if (this.options.interactionType === InteractionType.Popup) {
+					const response: AuthenticationResult = await this.publicClientApplication.acquireTokenPopup({ scopes });
+					return response.accessToken;
+				}
```
Comment (baywet) [inline src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts:69]: maybe log if we get and unexpected value?

File: package.json (hunk 2 above)
Comment (nikithauc) [reply package.json:131]: #485 Separate PR for removing msal.

File: src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts (hunk 5 above)
Comment (nikithauc) [reply src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts]: Changed this

Comment (baywet) [reply src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts]: sorry, I don't think I was explicit enough: what if the error is not InteractionRequiredAuthError or the type is not Popup/Redirect? I think we should log/throw for those totally unexpected cases

File: docs/MSALBrowserAuthenticationProvider.md (hunk 1 above)
Comment (nikithauc) [reply docs/MSALBrowserAuthenticationProvider.md]: Added reference to the browser cdn usage for more information. The sample uses the basic usage of CDN for the source.

File: src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts (hunk 5 above)
Comment (nikithauc) [reply src/authentication/msal-browser/MSALBrowserAuthenticationProvider.ts]: Thanks for pointing out. Added an else statement to throw the error.

**Review by baywet — DISMISSED** [review]

**Review by baywet — APPROVED** [review]

**Pull request merged**